### PR TITLE
Add api to resolve components

### LINKED ISSUE #5413: SelectorComponents do not get resolved
### Expected behavior

Get a message that contains the full list of entities

### Observed/Actual behavior

Get message `@e`

### Steps/models to reproduce
```java
@Override
public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
    Component component = Component.selector("@e");
    sender.sendMessage(component);
}
```
### Plugin list

None, except the one I used to verify this behavior.

### Paper version

> version
[00:16:21 INFO]: Checking version, please wait...
[00:16:21 INFO]: This server is running Paper version git-Paper-567 (MC: 1.16.5) (Implementing API version 1.16.5-R0.1-SNAPSHOT)
You are running the latest version

### Agreements

- [X] I am running the latest version of Paper available from https://papermc.io/downloads.
- [X] I have searched for and ensured there isn't already an open issue regarding this.
- [X] My version of Minecraft is supported by Paper.

Creating and sending `SelectorComponent`s does currently not work properly as the selector does not get resolved. I haven't tested it but same problem might apply to all component types listed [here](https://minecraft.fandom.com/wiki/Raw_JSON_text_format#Component_resolution) (i.e. `score`, `selector`, and `nbt`)

### PR DESCRIPTION
**Pull request by Machine-Maker**
Resolves https://github.com/PaperMC/Paper/issues/5413

Perhaps some mention of the permission craftbukkit adds `minecraft.command.selector` since that will block the use of some components that make use of selectors if the CommandSender doesn't have that permission?

### CODE REVIEW

File: patches/api/0005-Adventure.patch [hunk 1]
```diff
@@ -383,6 +388,19 @@ index 0000000000000000000000000000000000000000..15ecb12fd2fefcac96edbaef7cdd487a
...
 +    /**
++     * Resolves a component with a specific command sender and subject
++     *
++     * @param input the component to resolve
++     * @param context the command sender to resolve with
++     * @param scoreboardSubject the scoreboard subject to use (for use with {@link net.kyori.adventure.text.ScoreComponent}s)
++     * @return the resolved component
++     * @throws IOException if a syntax error tripped during resolving
++     */
++    public static @NotNull Component resolve(@NotNull Component input, @Nullable CommandSender context, @Nullable Entity scoreboardSubject) throws IOException {
```
Comment (kashike) [inline patches/api/0005-Adventure.patch]:
```suggestion
+    public static @NotNull Component resolveWithContext(@NotNull Component input, @Nullable CommandSender context, @Nullable Entity scoreboardSubject) throws IOException {
```

File: patches/api/0005-Adventure.patch [hunk 2]
```diff
@@ -1114,6 +1132,7 @@ index 01e11f882abb6c631f810584aa23646042688435..f43ad19a44f3b7368a1a110cc5524727
 +    net.kyori.adventure.text.serializer.gson.GsonComponentSerializer gsonComponentSerializer();
 +    net.kyori.adventure.text.serializer.gson.GsonComponentSerializer colorDownsamplingGsonComponentSerializer();
 +    net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer legacyComponentSerializer();
++    net.kyori.adventure.text.Component resolve(net.kyori.adventure.text.Component component, org.bukkit.command.CommandSender context, org.bukkit.entity.Entity scoreboardSubject) throws java.io.IOException;
```
Comment (kashike) [inline patches/api/0005-Adventure.patch]:
```suggestion
+    net.kyori.adventure.text.Component resolveWithContext(net.kyori.adventure.text.Component component, org.bukkit.command.CommandSender context, org.bukkit.entity.Entity scoreboardSubject) throws java.io.IOException;
```

File: patches/api/0005-Adventure.patch (hunk 1 above)
Comment (Machine-Maker) [reply patches/api/0005-Adventure.patch]: renamed as such

**Review by kashike — APPROVED** [review]

**Comment by Machine-Maker** [comment]
Rebased for 1.19.1

**Review by Owen1212055 — APPROVED** [review]
Similar to the instrument PR, maybe this can also be moved to its own provider api.
ComponentResolverProvider

Something like that. Logic looks good tho.

**Comment by jpenilla** [comment]
The selectEntities method does something similar with creating a CommandSourceStack and it adds a boolean param to override permissions, might want to do the same here

**Comment by Machine-Maker** [comment]
> The selectEntities method does something similar with creating a CommandSourceStack and it adds a boolean param to override permissions, might want to do the same here

Ok did this, entity selector permission check is bypassed. 


I also expanded the javadocs to explain what components can be resolved, and what is required/optional for each type.

**Review by jpenilla — APPROVED** [review]
Are there situations where we would want to check permissions? Maybe a boolean param for the override would be good. Looks good otherwise

**Pull request merged**